src/core/send.c

```diff
@@ -114,16 +114,58 @@ QuicSendCanSendFlagsNow(
     return TRUE;
 }
 
+#pragma warning(push)
+#pragma warning(disable:6001) // SAL thinks Connection could be uninitialized?
 _IRQL_requires_max_(DISPATCH_LEVEL)
 void
 QuicSendQueueFlush(
     _In_ QUIC_SEND* Send,
     _In_ QUIC_SEND_FLUSH_REASON Reason
     )
 {
+    QUIC_CONNECTION* Connection = QuicSendGetConnection(Send);
+
+#ifdef QUIC_USE_RAW_DATAPATH
+
+    QUIC_PATH* Path = &Connection->Paths[0];
+    QUIC_STATUS Status;
+    if (Path->Route.RouteState == RouteUnresolved) {
```

Comment:
I feel we shouldn't wait until the send flush to queue the route resolution. Let's do it ASAP, in the receive path, the first time we see the path.

Reply:
Also, might be a good idea to wrap a lot of this with a `QuicPathResolveRoute` function.

Reply:
Premature optimization IMO.

Reply:
Nick, I don't understand your suggestion. In the receive path, the route is automatically resolved and we don't need to queue anything for route resolution.

Reply:
I don't remember we decided to not trust L2 info. Why can't we trust it? I think we should trust it and refresh it when we see signals indicating it can't be trusted (like persistent loss etc).

Reply:
@maolson-msft what are your recollections? I thought we talked about "Is it Ok to just use the L2 info when we receive a packet" and it was decided that the path back might be different.

Reply:
Yeah, the theory was that there could be asymmetrical routes where RX packets come from one gateway but TX packets go to a different gateway. Whether this is a likely scenario is questionable, but it's definitely a POSSIBLE scenario.

I really think we ought to, for starters, just do route resolution in the TX path. That's what we need to be functional and there are no questions about correctness.

Reply:
> > Why was the route resolution moved from SendDataAlloc to SendQueueFlush?
> 
> If we do it in SendDataAlloc, the design would be much more complicated. SendDataAlloc happens right before we call into datapath. If at this point, route is still in resolving state, we would need to queue up frames in datapath, which requires its own synchronization, without mac address filled. When route is resolved, then we would need to go over each queued frame to fill the mac address.
> 
> If we do it in SendQueueFlush, we don't have this problem because we build frames once we know what to fill in L2 headers.

OK, makes sense.

Reply:
Makes sense. I think I will need to change how the route in path is being overwritten and ignore the state variable.